std/algorithm/comparison.d

```diff
@@ -732,78 +731,133 @@ template equal(alias pred = "a == b")
+    bool equal(R, Ss...)(R r, Ss ss)
+    if (ss.length >= 1 &&
+        allSatisfy!(isInputRange, R, Ss) &&
+        (!isInfinite!R || !allSatisfy!(isInfinite, Ss)) // at least one must be finite
+        )
     {
-        static assert(!(isInfinite!Range1 && isInfinite!Range2),
-            "Both ranges are known to be infinite");
-
-        //No pred calls necessary
-        static if (isEmptyRange!Range1 || isEmptyRange!Range2)
-        {
-            return r1.empty && r2.empty;
-        }
-        else static if ((isInfinite!Range1 && hasFixedLength!Range2) ||
-            (hasFixedLength!Range1 && isInfinite!Range2))
-        {
-            return false;
-        }
-        //Detect default pred and compatible dynamic array
-        else static if (is(typeof(pred) == string) && pred == "a == b" &&
-            isArray!Range1 && isArray!Range2 && is(typeof(r1 == r2)))
-        {
-            return r1 == r2;
-        }
-        // if one of the arguments is a string and the other isn't, then auto-decoding
-        // can be avoided if they have the same ElementEncodingType
-        else static if (is(typeof(pred) == string) && pred == "a == b" &&
-            isAutodecodableString!Range1 != isAutodecodableString!Range2 &&
-            is(ElementEncodingType!Range1 == ElementEncodingType!Range2))
+        enum isEquableToR(T) = is(typeof({ return R.init == T.init; }));
+
+        // start by detecting default pred and compatible dynamic array
+        static if (is(typeof(pred) == string) &&
+                   pred == "a == b" &&
+                   isArray!R &&
+                   allSatisfy!(isArray, Ss) &&
+                   allSatisfy!(isEquableToR, Ss))
         {
-            import std.utf : byCodeUnit;
-
-            static if (isAutodecodableString!Range1)
-            {
-                return equal(r1.byCodeUnit, r2);
-            }
-            else
+            foreach (ref s; ss)
             {
-                return equal(r2.byCodeUnit, r1);
+                if (r != s) return false;
             }
+            return true;
         }
-        //Try a fast implementation when the ranges have comparable lengths
-        else static if (hasLength!Range1 && hasLength!Range2 && is(typeof(r1.length == r2.length)))
+        // use fast implementation when the ranges have comparable lengths
+        else static if (allSatisfy!(hasLength, R, Ss))
         {
-            immutable len1 = r1.length;
-            immutable len2 = r2.length;
-            if (len1 != len2) return false; //Short circuit return
-
-            //Lengths are the same, so we need to do an actual comparison
-            //Good news is we can squeeze out a bit of performance by not checking if r2 is empty
-            for (; !r1.empty; r1.popFront(), r2.popFront())
+            // check equal lengths
+            foreach (ref s; ss)
+            {
+                if (r.length != s.length) { return false; }
```

Comment:
This test should go like this: for all ranges that implement `length`, lengths must compare equal. Should be the first test in the function.

Reply:
Ah, I made this comment a while ago apparently.